Add option module to Belt

arnarthor opened the pull request:
This adds the following utility functions for the option data type to belt.

```ocaml
val getExn : 'a option -> 'a
val mapWithDefaultU : 'a option -> 'b -> ('a -> 'b [@bs]) -> 'b
val mapWithDefault : 'a option -> 'b -> ('a -> 'b) -> 'b
val mapU : 'a option -> ('a -> 'b [@bs]) -> 'b option
val map : 'a option -> ('a -> 'b) -> 'b option
val flatMapU : 'a option -> ('a -> 'b option [@bs]) -> 'b option
val flatMap : 'a option -> ('a -> 'b option) -> 'b option
val getWithDefault : 'a option -> 'a -> 'a
val isSome : 'a option -> bool
val isNone : 'a option -> bool
val eqU : 'a option -> 'b option -> ('a -> 'b -> bool [@bs]) -> bool
val eq : 'a option -> 'b option -> ('a -> 'b -> bool) -> bool
val cmpU : 'a option -> 'b option -> ('a -> 'b -> int [@bs]) -> int
val cmp : 'a option -> 'b option -> ('a -> 'b -> int) -> int
```

chenglou commented on jscomp/others/belt_Option.ml:
```diff
@@ -0,0 +1,27 @@
+let getExn = function
+  | Some x -> x
+  | None -> assert false
```
I believe @bobzhang uses `[%assert "getExn"]` for some reason

chenglou commented on jscomp/others/belt_Option.ml:
```diff
@@ -0,0 +1,27 @@
+  | Some x -> x
+  | None -> assert false
+
+let fold opt default f = match opt with
```
The name will need some thinking

arnarthor replied:
People from the JS world probably know this as `reduce` or something along those lines at least for Lists/Arrays. We use fold internally because we're used to it from Scala/Haskell.

chenglou commented on jscomp/others/belt_Option.ml:
```diff
@@ -0,0 +1,27 @@
+  | Some x -> f x
+  | None -> default
+
+let map opt f = match opt with
```
I think we'd also expose some `mapU` like other modules for consistency. But those can wait after this PR

chenglou commented on jscomp/others/belt_Option.ml:
```diff
@@ -0,0 +1,27 @@
+  | Some x -> Some (f x)
+  | None -> None
+
+let flatMap opt f = match opt with
```
Same with `fold` regarding naming

chenglou commented on jscomp/others/belt_Option.ml:
```diff
@@ -0,0 +1,27 @@
+  | Some x -> x
+  | None -> default
+
+let exists = function
```
Belt.List uses `has`. cc @bobzhang another thing we should unify around. `has` vs `contains` vs `exists`. First one is easier for non-english speakers

chenglou replied:
Actually, not sure why `Belt.List.has` exists

chenglou commented on jscomp/others/belt_Option.ml:
```diff
@@ -0,0 +1,27 @@
+  | Some _ -> true
+  | None -> false
+
+let empty = function
```
This should be `isEmpty`

arnarthor commented:
I've addressed most of your comments @chenglou. I haven't changed the names for fold and flatMap.

Added uncurried versions of all the functions where that applies.

bobzhang commented on jscomp/others/belt_Option.mli:
```diff
@@ -0,0 +1,39 @@
+val flatMapU : 'a option -> ('a -> 'b option [@bs]) -> 'b option
+val flatMap : 'a option -> ('a -> 'b option) -> 'b option
+val getOrElse : 'a option -> 'a -> 'a
+val has : 'a option -> bool
```
`has` is redundant?

bobzhang commented on jscomp/others/belt_Option.mli:
```diff
@@ -0,0 +1,39 @@
+val map : 'a option -> ('a -> 'b) -> 'b option
+val flatMapU : 'a option -> ('a -> 'b option [@bs]) -> 'b option
+val flatMap : 'a option -> ('a -> 'b option) -> 'b option
+val getOrElse : 'a option -> 'a -> 'a
```
`getWithDefault`?

bobzhang commented on jscomp/others/belt_Option.mli:
```diff
@@ -0,0 +1,39 @@
+val mapU : 'a option -> ('a -> 'b [@bs]) -> 'b option
+val map : 'a option -> ('a -> 'b) -> 'b option
+val flatMapU : 'a option -> ('a -> 'b option [@bs]) -> 'b option
+val flatMap : 'a option -> ('a -> 'b option) -> 'b option
```
`bind` or `andThen`?

cullophid replied:
Flatmap. It makes much more sense when you need to explain it to newcommers

bobzhang commented on jscomp/others/belt_Option.mli:
```diff
@@ -0,0 +1,39 @@
+
+val getExn : 'a option -> 'a
+val foldU : 'a option -> 'b -> ('a -> 'b [@bs]) -> 'b
+val fold : 'a option -> 'b -> ('a -> 'b) -> 'b
```
is `fold` common enough to be included?

arnarthor replied:
We at least use it pretty often internally when dealing with a CLI tool we're writing. It's nice to have it to set defaults for missing arguments

bobzhang replied:
this name does not sound meaningful to me, `mapWithDefault` looks better?

bobzhang commented on jscomp/others/belt_Option.mli:
```diff
@@ -0,0 +1,39 @@
+val flatMap : 'a option -> ('a -> 'b option) -> 'b option
+val getOrElse : 'a option -> 'a -> 'a
+val has : 'a option -> bool
+val isEmpty : 'a option -> bool
```
I kinda feel `x = None` looks shorter than `isEmpty` and more efficient.

Some missed functionalities, `eq`, `cmp`?

mchaver commented:
@arnarthor How about a `catSomes` functions
```
let catSomes (type a) (mas: (a option) list) :a list =
  List.fold_left(fun acc x ->
      match x with
      | None -> acc
      | Some a -> List.append acc [a]
    ) [] mas
```

and an array version

```
let arrayCatSomes (type a) (mas: (a option) array) :a array =
  Array.fold_left(fun acc x ->
      match x with
      | None -> acc
      | Some a -> Array.append acc [a]
    ) [||] mas
```

arnarthor commented:
I don't really see a usecase for it to be honest. This is just a specialized version of something like `keep/filter` for both Lists and Arrays right?

mchaver commented:
@arnarthor It's useful if you have collections of computations that can potentially fail, then you get the successful ones. It's a common function in the stdlibs of other functional programming languages like Haskell, Scala, Purescript, etc. `catMaybes` in those languages.

utkarshkukreti commented on jscomp/others/belt_Option.ml:
```diff
@@ -0,0 +1,58 @@
+  | None -> default
+
+let has = function
+  | Some _ -> true
```
How about naming these `isSome` and `isNone` like Rust? I think they convey what the function actually does better than `has`/`isEmpty`.

brentvatne replied:
agreed. In addition to `isSome` `isNone`:

`Option.unwrap(Some(A))` vs `Option.get(Some(A))`
`Option.unwrapOr(Some(A), B)` vs `Option.getOrElse(Some(A), B)` vs `Option.getWithDefault(Some(A), B)`

arnarthor replied:
`isSome` and `isNone` are nice, but @bobzhang pointed out that `x == None` is more performant so they actually might not be needed.

I'd say that we should keep `getExn` and `getWithDefault` as is for consistency with other Belt modules which have `get`, `getExn`, `getBy` etc.

brentvatne replied:
I agree with suffix consistency there but `Option.get` doesn't communicate to me that we are unwrapping / moving / extracting the value contained in the `Option`. `unwrapExn` `unwrapWithDefault` imo are more clear, but I don't feel super strongly about it

arnarthor commented:
@bobzhang I've added `cmp` and `eq` functions. I haven't renamed `flatMap` yet since we there isn't really a clear decisions on what should be used instead. I personally feel like `flatMap` is the most descriptive in this case, but I really don't mind if it's `flatMap` or `bind`. I don't like `andThen` in this case since it's not really descriptive for options.

I also kept `has` and `isEmpty` there but renamed them to `isSome` and `isNone`. I know they're redundant since you could use `x === None` but I feel like having these verbose functions could help newcomers read through code.

Anyways feel free to leave some more comments if you would like me to change something 👍

bobzhang commented:
Lets see if we can make it for release, see my comments about the naming of `fold`, is `fold` a good name for that function

chenglou commented:
Yeah I think `mapWithDefault` is a better name for `fold` too

cullophid commented:
Since List already has map and flatten it would kinda make sense to call the combination flatMap...

arnarthor commented:
Done @bobzhang. Think it's ready for merge 👍

arnarthor commented:
I can't run a rebuild on travis, the instance stalled which caused the error.

chenglou commented:
Restarted for you. Related: #2564

chenglou commented:
I'm iffy on the name flatMap/bind/andThen, since it causes so much drama. Maybe we should leave it out for one version?
